src/Disks/IO/ThreadPoolReader.cpp

```diff
@@ -113,7 +113,7 @@ std::future<IAsynchronousReader::Result> ThreadPoolReader::submit(Request reques
     /// disable pread with nowait.
     static const bool has_pread_nowait_support = !hasBugInPreadV2();
 
-    if (has_pread_nowait_support)
+    if (has_pread_nowait_support && !request.direct_io)
```

Comment:
```suggestion
    /// RWF_NOWAIT is ignored for O_DIRECT (mostly, it may return EAGAIN if it cannot lock the inode in case of ext4)
    if (has_pread_nowait_support && !request.direct_io)
```